### ISSUE TITLE
2.0.0a1 Multiline `SELECT` are not correctly indented in CTA

### ISSUE THREAD

**Issue by dmohns** [sqlfluff/sqlfluff#4234]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Consider the following two examples:

```sql
CREATE TABLE mytable
AS
(SELECT
    id,
    user_id
FROM another_table
)
;
```

This one ☝️ does not throw any lint errors which is wrong I think. The `SELECT` should be on a new, indented line.

Compare this to below example that uses only a single SELECT target.

```sql
CREATE TABLE mytable
AS
(SELECT id
FROM table
)
;
```

which gets linted to

```sql
CREATE TABLE mytable
AS
(
    SELECT id
    FROM table
)
;
```

### Expected Behaviour

see above

### Observed Behaviour

see above

### How to reproduce

see above

### Dialect

ansi

### Version

2.0.0a1

### Configuration

default

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by alanmcruickshank** [comment]
Good catch! This is a very strange one - I'll investigate

**Comment by alanmcruickshank** [comment]
Ok - this is a potentially hairy one.

We register the first potential indent as untaken (between `(` and `SELECT`), but by a quirk of the logic, in the case where there isn't a newline before we get to the one before `FROM` it's deemed safe to insert a newline. In the other case though it's not because we don't yet know that the dedent _is_ taken yet.

Resolving this is going to be hard without introducing more passes through the file or by forcing every untaken indent (which I think is too aggressive).

I'm going to come back to this one once we've got some more bug fixed.

**Comment by fmms** [comment]
To my understanding the issue here is a lot wider than just CTAs (`CREATE TABLE AS`), it applies to CTEs  (common table expressions):

Have a look at the following example:
```sql
WITH m AS (SELECT
  firstCol
  , secondCol
FROM dbo.myTable
)

SELECT * FROM m
```

**Comment by alanmcruickshank** [comment]
I think I have a solution to this. Noting it down here so that I don't forget, although I won't be able to pick this up straight away. Ideally I'll get the rules reorg and format command first and then come back to this.

The problem here is that we allow indents to be _untaken_. Normally this is fine, but in this case the indent is untaken - but it's corresponding dedent _is taken_. There are two ways to catch this while still avoiding looping through the file to _search_ for whether the dedent is taken or not:
1. Force indents to always be taken. I don't like this solution and I think it would result in some very over-indented SQL.
2. Match dedents against their indents and catch whether they're taken or not. Force any indents which have corresponding dedents which _are_ taken. This approach is impossible to do in only one pass through the file - *but we already do two passes*: one to crawl the file and summarise the information and the second to detect and apply the fixes. We would do the matching in the first pass - store the information and then check against it in the second pass.

I think the second route is the only viable one. Edits would involve:
- modifications to the crawl routines at the start of the reindent process.
- additional clauses in the "missing positive indent" code to force indents the places detected.

I'll implement this in due course. Ideally pre 2.0.0 - but worse case as a fast follow in 2.0.1.

**Comment by dmohns** [comment]
It would be great if we can have this is part of the `2.0.0` release as this will potentially reformat a lot of queries (at least in my code base). So, shipping it later would require to do the same reformatting of queries again as will the initial `2.0.0` release. 

But, of course, it wouldn't make sense to delay the release __just__ because of this issue 😉
